### PR TITLE
use entitycache for list endpoints

### PR DESCRIPTION
**Pull request by rogpeppe**
We make the entitycache package API slightly more general
so that it can deal with the output of the list pipe.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1447/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1448/
Test PASSed.

File: internal/charmstore/store.go [hunk 1]
```diff
@@ -1521,8 +1521,12 @@ var sortMongoFields = map[string]string{
 
 // createSort creates a sort query parameters for mongo out of a Sort parameter.
```
Comment (mhilton) [inline internal/charmstore/store.go]: createMongoSort

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -1537,60 +1541,73 @@ func createMongoSort(sp SearchParams) (bson.D, error) {
 	return sort, nil
 }
 
-// List lists the store for the given ListParams.
-// It returns a ListResult containing the results of the list.
-func (store *Store) List(sp SearchParams) (ListResult, error) {
+// Listquery holds a list query from which an iterator
```
Comment (mhilton) [inline internal/charmstore/store.go]: ListQuery

**Comment by mhilton** [comment]
:+1:

**Comment by fabricematrat** [comment]
:+1:

File: internal/charmstore/store.go (hunk 1 above)
Comment (rogpeppe) [inline internal/charmstore/store.go]: done

File: internal/charmstore/store.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store.go]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1449/
Test PASSed.

**Pull request merged**